### PR TITLE
Add addr pool in rule

### PR DESCRIPTION
**Pull request by NikitaUnisikhin**
-- Added address to the rule. Address supported IPv4/IPv6 with mask and hostname
-- Сhanged the logs taking into account the new parameter
-- Сreated a module for working with addresses (address.c) also used the module in hba
-- Added tests for new functionality

### CODE REVIEW

File: sources/address.c [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+			if (ch2 >= 'A' && ch2 <= 'Z')
+				ch2 += 'a' - 'A';
+			else if (IS_HIGHBIT_SET(ch2) && isupper(ch2))
+				ch2 = tolower(ch2);
+
+			if (ch1 != ch2)
+				return (int) ch1 - (int) ch2;
+		}
+		if (ch1 == 0)
+			break;
```
Comment (AndrewOvvv) [inline sources/address.c:152]: Почему мы считаем, что s1 короче s2? (типа, почему мы только ch1 проверяем)

File: sources/address.c [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+	/* Lookup IP from host name and check against original IP */
+	ret = getaddrinfo(client_hostname, NULL, NULL, &gai_result);
+	if (ret != 0)
+		return false;
+
+	found = false;
+	for (gai = gai_result; gai; gai = gai->ai_next)
+	{
+		if (gai->ai_addr->sa_family == client_sa->ss_family)
```
Comment (AndrewOvvv) [inline sources/address.c:202]: Не суверен, но разве od_address_equals не делает всё то, что  происходит внутри?

File: sources/address.c [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+	char *end = NULL;
+	long len = strtoul(prefix, &end, 10);
+	if (*prefix == '\0' || *end != '\0') {
+		return -1;
+	}
+	if (address_range->addr.ss_family == AF_INET) {
+		if (len > 32)
+			return -1;
+		struct sockaddr_in *addr = (struct sockaddr_in *)&address_range->mask;
+		long mask;
```
[the remark this answers is not in the document]
Comment (AndrewOvvv) [reply sources/address.c:42]: I think that it is not good idea to used signed type for mask. There're exist some UB moments with signed integer types and bitwise operations

File: sources/address.c [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+		return true;
+	}
+
+	return false;
+}
+
+bool od_address_range_equals(od_address_range_t *first, od_address_range_t *second)
+{
+	if (first->is_hostname == second->is_hostname)
+		return first->string_value == second->string_value;
```
Comment (x4m) [inline sources/address.c:123]: od_address_strcasecmp()?

File: sources/address.c (hunk 1 above)
Comment (x4m) [reply sources/address.c:152]: It's expected that ch1==ch2 at this point...
but, idk, there's strcasecmp() https://pubs.opengroup.org/onlinepubs/9699919799/functions/strcasecmp.html

File: sources/address.c [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+	if (reti)
+		return -1;
+	reti = regexec(&regex, hostname, 0, NULL, 0);
+	if (reti == 0)
+		return 0;
+	else
+		return 1;
+}
+
+uint32 od_address_bswap32(uint32 x)
```
Comment (x4m) [inline sources/address.c:277]: BTW we have od_hba_bswap32().
But I'd prefer to have od_bswap32() once and for all :)

File: sources/router.c [hunk 6]
```diff
@@ -348,14 +349,24 @@ od_router_status_t od_router_route(od_router_t *router, od_client_t *client)
 
 	/* match latest version of route rule */
 	od_rule_t *rule;
+
+	struct sockaddr_storage sa;
+	if (!client->is_watchdog) {
```
Comment (x4m) [inline sources/router.c:354]: Why do we exempt watchdog? Can we exampt all outgoing clients, not just watchdog?

File: sources/address.c [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+{
+	od_address_range_t address_range = {
+		.string_value = strdup("all"),
+		.string_value_len = strlen("all"),
+		.is_default = 1
+	};
+	return address_range;
+}
+
+int od_address_range_copy(od_address_range_t *src, od_address_range_t *dst)
```
Comment (x4m) [inline sources/address.c:21]: maybe free() hostname?

File: sources/rules.c [hunk 8]
```diff
@@ -178,6 +178,8 @@ void od_rules_rule_free(od_rule_t *rule)
 		free(rule->db_name);
 	if (rule->user_name)
 		free(rule->user_name);
+	if (rule->address_range.string_value)
```
Comment (x4m) [inline sources/rules.c:181]: maybe free address_range too?

**Review by x4m — CHANGES_REQUESTED** [review]
Generally looks good to me, however feels very leaky. Let's make sure all allocations are free()ed correctly.

File: sources/address.c (hunk 1 above)
Comment (NikitaUnisikhin) [reply sources/address.c:152]: this case is considered in the next iteration of the loop

File: sources/address.c [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+	dst->addr = src->addr;
+	dst->mask = src->mask;
+	dst->is_default = src->is_default;
+	dst->is_hostname = src->is_hostname;
+}
+
+int od_address_range_read_prefix(od_address_range_t *address_range, char *prefix)
+{
+	char *end = NULL;
+	long len = strtoul(prefix, &end, 10);
```
[the remark this answers is not in the document]
Comment (NikitaUnisikhin) [reply sources/address.c:34]: length can be <= zero (str 55)

[the remark this answers is not in the document]
Comment (AndrewOvvv) [reply sources/address.c:34]: but why do you use strtoul (to unsigned long)?

[the remark this answers is not in the document]
Comment (NikitaUnisikhin) [reply sources/address.c:34]: oh yeah, thanks for the comment, I'll fix it

File: sources/address.c [hunk 10]
```diff
@@ -28,28 +26,31 @@ int od_address_range_copy(od_address_range_t *src, od_address_range_t *dst)
...
 	if (address_range->addr.ss_family == AF_INET) {
 		if (len > 32)
 			return -1;
-		struct sockaddr_in *addr = (struct sockaddr_in *)&address_range->mask;
-		long mask;
+		struct sockaddr_in *addr =
+			(struct sockaddr_in *)&address_range->mask;
+		uint32 mask;
 		if (len > 0)
 			mask = (0xffffffffUL << (32 - (int)len)) & 0xffffffffUL;
```
Comment (AndrewOvvv) [inline sources/address.c:44]: it seems that you now don't need to do "& 0xffffffffUL" here

File: sources/address.c (hunk 2 above)
Comment (NikitaUnisikhin) [reply sources/address.c]: thanks, fixed

File: sources/address.c (hunk 4 above)
Comment (NikitaUnisikhin) [reply sources/address.c]: thanks, fixed

File: sources/address.c (hunk 5 above)
Comment (NikitaUnisikhin) [reply sources/address.c]: thanks, done

File: sources/router.c (hunk 6 above)
Comment (NikitaUnisikhin) [reply sources/router.c]: thanks, now we exempt all internal clients

File: sources/address.c (hunk 10 above)
Comment (NikitaUnisikhin) [reply sources/address.c]: thanks, fixed

File: sources/rules.c (hunk 8 above)
Comment (NikitaUnisikhin) [reply sources/rules.c:181]: I don't use malloc

File: sources/address.c (hunk 7 above)
Comment (NikitaUnisikhin) [reply sources/address.c:19]: I don't use malloc

**Review by x4m — APPROVED** [review]

**Pull request merged**